prov/verbs/src/verbs_eq.c

```diff
@@ -990,3 +1098,40 @@ int fi_ibv_eq_open(struct fid_fabric *fabric, struct fi_eq_attr *attr,
 	return ret;
 }
 
+int fi_ibv_eq_attach_domain(struct fi_ibv_eq *eq, struct fi_ibv_domain *domain)
+{
+	int async_fd = domain->verbs->async_fd;
+	struct epoll_event event;
+	int ret;
+
+	ret = fi_fd_nonblock(async_fd);
+	if (ret)
+		return ret;
+
+	memset(&event, 0, sizeof(event));
+	event.events = EPOLLIN;
+	event.data.ptr = domain;
+
+	if (epoll_ctl(eq->epfd, EPOLL_CTL_ADD, async_fd, &event))
+		return -errno;
+
+	fastlock_acquire(&eq->lock);
+	dlist_insert_tail(&domain->eq_dlentry, &eq->domain_list);
+	fastlock_release(&eq->lock);
+	return 0;
+}
+
+int fi_ibv_eq_dettach_domain(struct fi_ibv_eq *eq, struct fi_ibv_domain *domain)
+{
+	int async_fd = domain->verbs->async_fd;
+
+	/* Always remove the entry from the list */
+	fastlock_acquire(&eq->lock);
+	dlist_remove(&domain->eq_dlentry);
+	fastlock_release(&eq->lock);
+
+	if (epoll_ctl(eq->epfd, EPOLL_CTL_DEL, async_fd, NULL))
+		return -errno;
+
+	return 0;
+}
```

Comment:
I would expect the attach / detach calls to modify a reference count on the EQ, to ensure that it is non-zero when being destroyed.